### PR TITLE
Checkpoint Query Improvements

### PR DESCRIPTION
**Pull request by WindOfRussia**
## Changes

- Improved query scopes execution plan
- updated relations across the board
- updateOrCreateRevision() simplified to startRevision()
- saveAsRevision() -> performRevision() : cleaned up
- Removed created_at handling
- added deleted observer that preserves revision history on forceDeletes
- added dynamic relations for initial / previous / next revisions of a model (via select subqueries)
- added tests (still missing tests for Excluded / Unwatched, Revisions Model and more coverage on HasRevisions)
- cast metadata as array in php land, added a convenience scope for fetching revisionables directly with metadata
- reorganized the config file
- updated php-unit to 9.3
- added psalm static analysis tool
- added GitHub actions to run tests

### CODE REVIEW

File: src/Concerns/HasRevisions.php
Comment (m-triassi) [inline src/Concerns/HasRevisions.php:159]: Something about this seems off, but i cant put my finger on what. Can you explain the rational behind this one?

File: src/Models/Revision.php
Comment (m-triassi) [inline src/Models/Revision.php:115]: Was originally going to ask why you removed the json_decode from the meta mutator, until i saw this. Now i have a new question though, don't we have JSON stored in the metadata field, or am i mis-remembering?

File: src/Models/Revision.php
Comment (m-triassi) [inline src/Models/Revision.php:126]: hah! what were we thinking 😂 good catch

File: src/Models/Revision.php
Comment (m-triassi) [inline src/Models/Revision.php:237]: Doesn't removing this mean we lose the ability to pass in strings like "tomorrow"? i know that probably wont come up super often, but seems like it could be handy, no?

File: src/Observers/RevisionableObserver.php
Comment (m-triassi) [inline src/Observers/RevisionableObserver.php:106]: Just noting that this might be superseded by the changes coming to fix #2, more as a note to myself

File: src/Scopes/RevisionScope.php
Comment (m-triassi) [inline src/Scopes/RevisionScope.php:67]: We should probably find a better place to put this stuff, maybe in a wiki page or something, under methodology or something like that? just a cosmetic comment though

File: src/Models/Revision.php
Comment (m-triassi) [reply src/Models/Revision.php:115]: Actually circling back to this after testing, it looks like the metadata gets stored as `"[]"` / interpreted as a string; can you check that out and see if you're getting the same thing? 
ie: `Post::withoutGlobalScopes()->find(2)->metadata` where id 2 is a post that was just revisioned gives me that string output from above, and not an empty array like id expect

**Review by m-triassi — CHANGES_REQUESTED** [review]
The only thing, id say blocking this from approval is the meta data comment on lines 113-115 on `Revision.php` everything else is largely cosmetic.

File: src/Concerns/HasRevisions.php
Comment (WindOfRussia) [reply src/Concerns/HasRevisions.php:159]: previous_id corresponds to the previous id of the current revisionable item. 
The next dynamic relation scopes (withPrevious, withNext, etc.) automatically add these keys to the attributes of the model, but in case those scopes are not applied I fallback to just fetching it from the relation chain.

File: src/Models/Revision.php
Comment (WindOfRussia) [reply src/Models/Revision.php:115]: Maybe we should add some tests for the metadata storage trait. I originally assumed that by casting it to array I would essentially skip the json_decode step. Let me look into this one.

File: src/Models/Revision.php
Comment (WindOfRussia) [reply src/Models/Revision.php:237]: I removed it to avoid the errors Carbon::parse throws when it receives strings that don't conform to `strtotime()`, ie the checkpoint slugs. Besides I think in this case, less is more. We already provide 3 ways to interpret these variables and a user could just run `Revisionable::at(strtotime("tomorrow"))->get()` to achieve the same result.

File: src/Scopes/RevisionScope.php
Comment (WindOfRussia) [reply src/Scopes/RevisionScope.php:67]: fair point. I'm just leaving it there for the sake of it existing somewhere. Given the column name clashes that can occur with joins I think its fair to assume that we'll stick to the whereHas query. 

Unrelated to your comment, but related to this section of code:
I wanted to leverage the join to automatically select the metadata column on every revisionable row but this will force us to use fully qualified column names in where queries, etc. Sadly that approach just creates more trouble down the line.

File: src/Concerns/HasRevisions.php
Comment (m-triassi) [reply src/Concerns/HasRevisions.php:159]: Ahhh yes ok i see; that makes sense, carry on here

File: src/Models/Revision.php
Comment (m-triassi) [reply src/Models/Revision.php:237]: Yeah good point there, no problem!

File: src/Scopes/RevisionScope.php
Comment (m-triassi) [reply src/Scopes/RevisionScope.php:67]: I dont really see a problem with using fully qualified column names besides being rather cumbersome

Comment (WindOfRussia) [reply src/Scopes/RevisionScope.php:67]: right, but it's not negligible workflow change, for not only ourselves but anyone using this package. 
Also renders the laravel magic `where<attribute>()` magic methods useless (`Post::whereSlug('aaa')->whereStatus('published')->etc...`)

we can always come back and change it back if we really want that join on revision.

Comment (m-triassi) [reply src/Scopes/RevisionScope.php:67]: Right, i hadn't considered that to be honest. I'll defer to your good judgement on that one!

File: src/Models/Revision.php
Comment (WindOfRussia) [reply src/Models/Revision.php:116]: @m-triassi should be resolved now.. also has a test to accompany it

Comment (m-triassi) [reply src/Models/Revision.php:116]: Good stuff looks like it was resolved by 106824293cfaf3c0e01661bcd03aa0bd49fee6de!

**Review by m-triassi — APPROVED** [review]
Looks like everything else added besides the stuff i've already reviewed are tests. They all pass and seem sane to me. Other stuff seems to be housekeeping, thank you for that <3. Did some manual testing, all seems to work as intended. Great stuff!

**Comment by m-triassi** [comment]
Just an aside, we probably want to look at the `excluded` vs `unwatched` and how we're handling that, could probably be handled via another PR though since that's separate from this. Merging!

**Pull request merged**
